playback/src/audio_backend/alsa.rs

```diff
@@ -54,28 +55,32 @@ fn open_device(dev_name: &str, format: AudioFormat) -> Result<(PCM, Frames), Box
         AudioFormat::S24_3 => Format::S243BE,
     };
 
-    // http://www.linuxjournal.com/article/6735?page=0,1#N0x19ab2890.0x19ba78d8
-    // latency = period_size * periods / (rate * bytes_per_frame)
-    // For stereo samples encoded as 32-bit float, one frame has a length of eight bytes.
-    let mut period_size = ((SAMPLES_PER_SECOND * format.size() as u32) as f32
-        * (BUFFERED_LATENCY.as_secs_f32() / BUFFERED_PERIODS as f32))
-        as Frames;
-    {
+    let bytes_per_period = {
         let hwp = HwParams::any(&pcm)?;
         hwp.set_access(Access::RWInterleaved)?;
         hwp.set_format(alsa_format)?;
         hwp.set_rate(SAMPLE_RATE, ValueOr::Nearest)?;
         hwp.set_channels(NUM_CHANNELS as u32)?;
-        period_size = hwp.set_period_size_near(period_size, ValueOr::Greater)?;
-        hwp.set_buffer_size_near(period_size * BUFFERED_PERIODS)?;
+        // Deal strictly in time and periods.
+        hwp.set_periods(NUM_PERIODS, ValueOr::Nearest)?;
+        hwp.set_period_time_near(PERIOD_TIME.as_micros() as u32, ValueOr::Nearest)?;
         pcm.hw_params(&hwp)?;
 
         let swp = pcm.sw_params_current()?;
-        swp.set_start_threshold(hwp.get_buffer_size()? - hwp.get_period_size()?)?;
+        // Don't assume we got what we wanted.
+        // Ask to make sure.
+        let frames_per_period = hwp.get_period_size()?;
+
+        swp.set_start_threshold(hwp.get_buffer_size()? - frames_per_period)?;
         pcm.sw_params(&swp)?;
-    }
+        // Our goal is to get the period size in bytes.
+        // Assuming we got what we wanted, a 125ms period:
+        // S16 44100 stereo:
+        // 5513                   * 2                             * 2 = 22052 bytes
+        (frames_per_period as u32 * NUM_CHANNELS as u32) as usize * format.size()
```

Comment:
One more from me: you could alternatively use `pcm.frames_to_bytes(frames_per_period)` here and let ALSA do the work.

Reply:
Brilliant, that works perfectly. No math is good math,lol!!!